### PR TITLE
Remove ignite

### PR DESCRIPTION
**Pull request by szympie**

### CODE REVIEW

File: brats/transformations.py [hunk 1]
```diff
@@ -85,20 +86,22 @@ def hist_match(source: np.ndarray, template: np.ndarray) -> np.ndarray:
...
+H = 0
+W = 1
+D = 2
+C = 3
+
 
 @reorder.register(np.ndarray)
 def _(img: np.ndarray) -> np.ndarray:
-    transformed = np.moveaxis(img, [H,W,D,C], [2,3,1,0])
+    transformed = np.moveaxis(img, [H, W, D, C], [2, 3, 1, 0])
```
Comment (myychal) [inline brats/transformations.py]: <3

File: brats/transformations.py [hunk 2]
```diff
@@ -200,3 +203,46 @@ def __call__(self, tensor: torch.Tensor) -> torch.Tensor:
...
+
+
+@one_hot_encoding.register(np.ndarray)
+def _(img: np.ndarray, classes: typing.List[int]) -> np.ndarray:
+    if 0 in classes:
+        classes.remove(0)  # Without the background label
+    new_shape = [len(classes)] + list(img.shape[1:])
+    transformed = np.zeros(new_shape)
+    for class_id, label in enumerate(filter(lambda label: label != 0, classes)):
+        transformed[class_id][img[0, ...] == label] = 1
```
Comment (myychal) [inline brats/transformations.py]: This 0 is very confusing to me, what does it represent?

File: train_unet.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                                          trfs.Lambda(lambda x: F.interpolate(x, size=args.input_size)),
+                                          trfs.Lambda(lambda x: x.float())
+                                          ])
+
+    volumes_paths, masks_paths = read_dataset_json(args.dataset_json)
+    volumes_set = datasets.NiftiFolder(volumes_paths, volumes_transformations)
+    masks_set = datasets.NiftiFolder(masks_paths, masks_transformations)
+    combined_set = datasets.CombinedDataset(volumes_set, masks_set)
+    with open(args.division_json, "r") as division_file:
+        indeces = json.load(division_file)
```
Comment (myychal) [inline train_unet.py]: typo in indeces

File: train_unet.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    train_set = Subset(combined_set, indeces["train"])
+    valid_set = Subset(combined_set, indeces["valid"])
+
+    train_loader = torch.utils.data.DataLoader(train_set, batch_size=args.batch_size, shuffle=True)
+    valid_loader = torch.utils.data.DataLoader(valid_set, batch_size=args.batch_size, shuffle=True)
```
Comment (myychal) [inline train_unet.py]: shuffle in valid kind of unnecessary

File: train_unet.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    model = UNet3D(1, 3).float()
+    model.to(args.device)
+    criterion = DiceLoss()
+    optimizer = optim.Adam(model.parameters(), args.learning_rate)
+    dice = DiceScore()
+    metrics = {"dice_edema": metric_for_class(dice, 0),
+               "dice_non_enhancing": metric_for_class(dice, 1),
+               "dice_enhancing": metric_for_class(dice, 2),
+               }
```
Comment (myychal) [inline train_unet.py]: Does it mean that the dice score will be calculated 3 times, for each class separately, even though one call gives individual score for each class?

File: train_unet.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        train_loss, train_metrics = run_training_epoch(model, train_loader, optimizer, criterion, metrics, args.device)
+        valid_loss, valid_metrics = run_validation_epoch(model, valid_loader, criterion, metrics, args.device)
+        print(f"Epoch: {epoch} "
+              f"Train loss: {train_loss:.4f} "
+              f"Valid loss: {valid_loss:.4f} "
+              f"Valid dice edema: {valid_metrics['dice_edema']:.4f} "
+              f"Valid dice non enhancing: {valid_metrics['dice_non_enhancing']:.4f} "
+              f"Valid dice enhancing: {valid_metrics['dice_enhancing']:.4f} ", flush=True)
+
+        last_model_logger.log(model, 0)
```
Comment (myychal) [inline train_unet.py]: Why always 0th epoch?

File: train_unet.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        parser.add_argument('--division_json', type=str, required=True)
+        parser.add_argument('--log_dir', type=str, required=True)
+        parser.add_argument('--device', type=str, required=True)
+        parser.add_argument('--epochs', type=int, required=True)
+        parser.add_argument('--batch_size', type=int)
+        parser.add_argument('--patience', type=int, default=10)
+        parser.add_argument('--learning_rate', type=float, default=0.001)
+        parser.add_argument('--train_valid_ratio', type=float, default=0.8)
+        parser.add_argument('--input_size', type=int, default=240)
+        parser.add_argument('--progress_bar', type=int, default=0)
```
Comment (myychal) [inline train_unet.py]: `train_val_ratio` and `progress_bar`  are unused

File: brats/transformations.py (hunk 1 above)
Comment (szympie) [reply brats/transformations.py]: Jedna literka o jak by było =]

File: train_unet.py (hunk 3 above)
Comment (szympie) [reply train_unet.py]: Where is the typo?

File: train_unet.py (hunk 6 above)
Comment (szympie) [reply train_unet.py]: This is sketchy, I want to cut the number of artifacts produced and I saved it always with the index 0. I changed the implementation to take suffix rather than epoch number.

Comment (szympie) [reply train_unet.py]: Fixed

File: train_unet.py (hunk 7 above)
Comment (szympie) [reply train_unet.py]: Fixed

File: train_unet.py (hunk 4 above)
Comment (szympie) [reply train_unet.py]: Changed

File: train_unet.py (hunk 3 above)
Comment (myychal) [reply train_unet.py]: I guess you mean plural of `index`, so it's `indices`

Comment (szympie) [reply train_unet.py]: Changed

File: brats/transformations.py (hunk 2 above)
Comment (szympie) [reply brats/transformations.py]: OK, this is connected to the indexing, I need to create a tensor with [N, D, H, W] shape, where N is number classes. So I do it and then want to fill it with the mask from the original one. 

If I would do
```python

transformed[class_id][mask == label] = 1

```
or
```python

transformed[mask == label] = 1

```
I would get index error, because there is a mismatch in shape. 
In first version, there would be too much indices in `mask` to properly index `transformed`.
In the second one, there would be correct number of indices, however we would try to index transformed (shape `[3, 155, 240, 240]`) with mask (shape `[1, 155, 240, 240]`). 

So what I do is to index each channel of `transformed` (`transformed[class_id]`) separately by squeezed version of `mask` (`mask[0, ...]`).

If you have a better solution, I would love to change that part of code as I really needed to focus in order to remember what I did here.

File: train_unet.py (hunk 5 above)
Comment (szympie) [reply train_unet.py]: i change the way the metrics are saved.

File: brats/transformations.py (hunk 2 above)
Comment (myychal) [reply brats/transformations.py]: Oh, I thought that the input parameter `img` has only three dimensions (D, H, W), but we add an additional one in the previous transformations. This was necessary in the case of binary classification, but do we need to keep that transformation which adds an aditional dimension in the case of multilcass, having this transformation? I guess yes, if we ever would want to go back to binary classification smoothly.

File: train_unet.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                                            trfs.Lambda(lambda x: x[3, :, :, :]),
+                                            trfs.Lambda(lambda x: np.expand_dims(x, 0)),
+                                            trfs.Lambda(lambda x: torch.from_numpy(x)),
+                                            trfs.Lambda(
+                                                lambda x: F.pad(x, [0, 0, 0, 0, 5, 0]) if x.shape[1] % 2 != 0 else x),
+                                            trfs.Lambda(lambda x: F.interpolate(x, size=args.input_size)),
+                                            transformations.StandardizeVolume(),
+                                            trfs.Lambda(lambda x: x.float()),
+                                            ])
+    masks_transformations = trfs.Compose([trfs.Lambda(lambda x: np.expand_dims(x, 3)),
```
Comment (myychal) [inline train_unet.py]: I found a transformation for this in `transformations.py` -> `AddChannelDimToMaskTransformation`. The question is whether we want to keep it here (see a few comments above)

File: brats/training/runners.py [hunk 9]
```diff
@@ -33,8 +33,9 @@ def run_training_epoch(model: nn.Module, data_loader: data.DataLoader, optimizer
         optimizer.step()
         losses.append(loss.item())
         for metric_name in metrics:
-            out_metrics[metric_name].append(metrics[metric_name](output, target).item())
-    out_metrics = {metric_name: np.mean(out_metrics[metric_name]) for metric_name in metrics.keys()}
+            metric_values = [value.item() for value in metrics[metric_name](output, target)]
```
Comment (myychal) [inline brats/training/runners.py]: This is great, but if I understand correcltly, it will fail if a metric returns a scalar value.  The question is: do we care?

File: brats/transformations.py (hunk 2 above)
Comment (szympie) [reply brats/transformations.py]: In case of od DL, I prefer to keep the channel dimension - I believe it is a way to emphasize that this channel is there. Of course, additionally going back to binary is convenient this way.

File: train_unet.py (hunk 8 above)
Comment (szympie) [reply train_unet.py]: I forgot about this one, changed.

File: brats/training/runners.py (hunk 9 above)
Comment (szympie) [reply brats/training/runners.py]: changed how metrics are calculated

File: train_unet.py (hunk 4 above)
Comment (myychal) [reply train_unet.py]: It's actually not

File: brats/training/runners.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            input = input.to(device)
+            target = target.to(device)
+            output = model(input)
+            loss = criterion(output, target)
+            losses.append(loss.item())
+            for metric_name in metrics:
+                metric_values = [value.item() for value in metrics[metric_name](output, target)]
+                out_metrics[metric_name].append(metric_values)
+        out_metrics = {metric_name: np.array(out_metrics[metric_name]).mean(axis=1) for metric_name in metrics.keys()}
+    return (np.mean(losses), out_metrics) if losses else (0.0, out_metrics)
```
Comment (myychal) [inline brats/training/runners.py]: When losses could ever be empty?

File: brats/metrics.py [hunk 11]
```diff
@@ -139,3 +140,16 @@ def __call__(self, prediction: torch.Tensor,
                     Dimensions - (Batch, Depth)
             """
         return F.hausdorff95(prediction, target, self.merge_operation)
+
+
+def metric_for_class(metric: typing.Callable[[torch.Tensor, torch.Tensor], torch.Tensor], class_id: int):
```
[the remark this answers is not in the document]
Comment (szympie) [reply brats/metrics.py]: Yeah, I left it there because I thought it would be useful in the future, but it is so simple to write that maybe it is better to delete it.

File: brats/training/loggers.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        """
+        Logging method
+        Args:
+            :param score: Metric, if it is improved, the state dict is saved (The higher the better).
+        """
+        if score > self.best_score:
+            self.best_score = score
+            model_path = os.path.join(self.log_directory, f"best_state_dict")
+            with open(model_path, "wb") as file:
+                torch.save(model, file)
```
[the remark this answers is not in the document]
Comment (szympie) [reply brats/training/loggers.py]: Fixed

File: brats/training/runners.py (hunk 10 above)
Comment (szympie) [reply brats/training/runners.py]: Yeah, it was something left from the previous project. Haven't noticed it
